### ISSUE TITLE
Re-use `proofValue` from DataIntegrityProof instead of `jws`

### ISSUE THREAD

**Issue by msporny** [w3c/vc-jws-2020#25]
We should discuss whether it's better to use `proofValue` from DataIntegrityProof instead of `jws`. There are upsides and downsides to both approaches. We might also explore if `JsonWebSignature2020` should be renamed to `JsonWebSignature` or `DataIntegrityProof`.

**Comment by mprorock** [comment]
> We might also explore if `JsonWebSignature2020` should be renamed to `JsonWebSignature` or `DataIntegrityProof`

+1

**Comment by msporny** [comment]
Here are a few options. 

The first is an attempt to re-use DataIntegrityProof, which would require encoding some sort of flattened JWS JSON Serialization into `proofValue` and prepending a multibase `u` value for base64url encoding. Upside there is that it fits nicely into the `DataIntegrityProof` format and benefits from CBOR-LD compression (if that's viewed as a benefit). Downside is that it is not 1) a detached signature, which is what implementations do today, 2) uses a `u` prefix, so it's not a raw output from a JWS library. 

```javascript
{
  "type": "DataIntegrityProof",
  "cryptosuite": "jws-2022",
  "created": "2022-11-29T20:35:38Z",
  "verificationMethod": "did:example:123456789abcdefghi#keys-1",
  "proofPurpose": "assertionMethod",
  "proofValue": "uDtEhU3ljbEg8L38VWAfUAqOyKAM6-Xx-F4GawxaepmXFCgfTjDxw5djxLa8ISlSApmWQxfKTUJqPP3-Kg6NU1Q"
}
```

The second option is to try and add a multibase encoding `e` for "JOSE-encoded dot notation", re-use `proofValue`, which feels like a horrible abuse of the multibase registry. The upside is that we can largely keep all the code the same, just change `jws` to `proofValue` and we're done, though you wouldn't get CBOR-LD compression w/o writing a JOSE compressor. Feels like an abuse of both multibase and CBOR-LD.

```javascript
{
  "type": "DataIntegrityProof",
  "cryptosuite": "jws-2022",
  "created": "2022-11-29T20:35:38Z",
  "verificationMethod": "did:example:123456789abcdefghi#keys-1",
  "proofPurpose": "assertionMethod",
  "proofValue": "eyJhbGciOiJFZERTQSIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..MJ5GwWRMsadCyLNXU_flgJtsS32584MydBxBuygps_cM0sbU3abTEOMyUvmLNcKOwOBE1MfDoB1_YY425W3sAg"
}
```

The third option is to loosen the restriction on `proofValue` to allow the use of `jws`, which might be the right balance to strike. We'd have to [loosen restrictions on `proofValue` here](https://w3c.github.io/vc-data-integrity/#dataintegrityproof): 

```javascript
{
  "type": "DataIntegrityProof",
  "cryptosuite": "jws-2022",
  "created": "2022-11-29T20:35:38Z",
  "verificationMethod": "did:example:123456789abcdefghi#keys-1",
  "proofPurpose": "assertionMethod",
  "jws": "eyJhbGciOiJFZERTQSIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..MJ5GwWRMsadCyLNXU_flgJtsS32584MydBxBuygps_cM0sbU3abTEOMyUvmLNcKOwOBE1MfDoB1_YY425W3sAg"
}
```

I'm leaning towards option 3, unless someone else can think of a better way of taking advantage of `proof` and `DataIntegrityProof` being defined in the base vc/v2 context? I haven't spent a lot of time thinking about this, so there might be a better approach.

/cc @dlongley

**Comment by OR13** [comment]
I've added a label to track changes that we would make based on data integrity, we can make them before it stabilizes, but I prefer to just wait and gather a larger group to do all at once.

**Comment by msporny** [comment]
> I've added a label to track changes that we would make based on data integrity, we can make them before it stabilizes, but I prefer to just wait and gather a larger group to do all at once.

As a related aside, I'm hoping to simplify the text that needs to be written in cryptosuite specs and will provide an example w/ the new `eddsa-2022` cryptosuite. That should make Editor's lives easier with the vc-jws-2022 cryptosuite if/when you apply some of the Data Integrity spec changes to it.

**Comment by F-Node-Karlsruhe** [comment]
Is there any outcome of this. We currently have the issue that without proper cryptosuite sec we are not able to determine the key type (ES256 or PS256) without fetching the public key itself.

What approach would you use for a production system right now. We are always signing as both DI-Proof and VC-JWT

Currently our DI-Proofs look like this

```
{
    "type": "JsonWebSignature2020",
    "created": "2025-12-05T08:19:45Z",
    "verificationMethod": "domain.com#test-key",
    "proofPurpose": "assertionMethod",
    "proofValue": "z5tTSQusJE9UfBjqg8rAoXbxJG19sheGev6cvc1pneT1wmB2t637HdpBjE1xKaxPufSH963Q83hHTYXwFhpSKndaz"
  }
```

**Comment by OR13** [comment]
If you must use data integrity proofs, I suggest using them the way they are defined here: https://www.w3.org/TR/vc-data-integrity/ .

In production systems I recommend to use JWS to sign things, because it is well supported in many languages and does not require normalization of attacker controlled inputs, see: https://www.w3.org/TR/rdf-canon/#dataset-poisoning

If someone wanted to resurrect this spec and make it work for JWS, you could modernize it and parameterize by algorithm, so instead of:

`eddsa-rdfc-2022`.... `jws-ES256-jcs-2025`, `jws-Ed25519-jcs-2025` (https://datatracker.ietf.org/doc/html/rfc9864#section-2.2)...

Note that eddsa by itself could lead to an ed448 key or an ed25519 key.

also... your `proofValue` is multibase... it should be a JWS, which contains the algorithm used to sign... https://w3c.github.io/vc-jws-2020/#example-6 (decoded headers):

```json
{
  "alg": "EdDSA",
  "b64": false,
  "crit": [
    "b64"
  ]
}
```

For ES256 with P-256, this would look like:

```json
{
  "alg": "ES256",
  "b64": false,
  "crit": [
    "b64"
  ]
}
```

But it would also have a proper JWS as the signature... not multibase stuff.

**Comment by F-Node-Karlsruhe** [comment]
Thanks for the comprehensive response, but i'm still a bit confused.

From my understanding using JWS  for DataIntegrityProof signing VCs with JWK would be the way to go but this method was deprecated/discontinued https://www.w3.org/TR/vc-jws-2020/

Therefore we went ahead and implemented it with proof value using JWKs like digitalbazaar does it https://github.com/digitalbazaar/ecdsa-multikey/blob/main/lib/index.js#L91 @dlongley in out signing package https://github.com/european-epc-competence-center/es256-signature-2020

This works as well but when verifying one has to go the extra step and get the "JWK header" from the DID document in order to extract the right crypto suite for verification. That is how we discovered about cryptosuite property which is from our understanding exactly the needed indicator for the suite to use for verification; but not a well defined one.

**Comment by msporny** [comment]
> That is how we discovered about cryptosuite property which is from our understanding exactly the needed indicator for the suite to use for verification; but not a well defined one.

Cryptosuite is defined here, in a standard: https://www.w3.org/TR/vc-data-integrity/#proofs

I took a brief look at https://github.com/european-epc-competence-center/es256-signature-2020

It seems like what you're doing is largely duplicative of an existing standard:

https://www.w3.org/TR/vc-di-ecdsa/

... which does support JWKs to express key material (also in a published standard): 

https://www.w3.org/TR/cid/#JsonWebKey

You seem to have at least two paths forward:

1. Re-use the vc-di-ecdsa standard for your purposes (which supports P-256 ECDSA signatures).
2. Define your own `es256-signature-2020` cryptosuite. I can't recommend doing this as it's not clear what significant benefit it would have over the existing vc-di-ecdsa standard (other than not canonicalizing at all, even with JCS, and re-using a JWS signature serialization as @OR13 noted, which are debatable benefits).

Have you considered using something like `ecdsa-jcs-2019` instead? https://www.w3.org/TR/vc-di-ecdsa/#ecdsa-jcs-2019

**Comment by OR13** [comment]
ES256 is shorthand for ECDSA with SHA256 and P-256.

In the context of JWS (with detached payload) it is described here:

- https://datatracker.ietf.org/doc/html/rfc7515
- https://datatracker.ietf.org/doc/html/rfc7797

If you are going to use vc data integrity, you might as well use option 1 above.

**Comment by F-Node-Karlsruhe** [comment]
Your remarks were really helpful :) We moved all data integrity signing to the digitalbazaar libraries, as well as the keys to multikey

https://github.com/european-epc-competence-center/key-service/pull/6

We also published the missing rsa-mutlikey package https://github.com/european-epc-competence-center/rsa-multikey  only to find out that something like rsa-rdfc-20xx does not yet exist. Will this spec be defined/supported soon? Afaiu. it is mostly about the canonization, which is the same for both https://github.com/digitalbazaar/ecdsa-rdfc-2019-cryptosuite/blob/main/lib/canonize.js and https://github.com/digitalbazaar/eddsa-rdfc-2022-cryptosuite/blob/main/lib/canonize.js , i.e. how the proof is generated

Is it save to apply a similar approach to RSA keys, e.g. in an own rsa-rdfc-2025-cryptosuite package as it is around the corner or is RSA out of scope for DataIntegrityProof?

**Comment by F-Node-Karlsruhe** [comment]
I just did it for now so we can offer DI Proof for RSA like for the others as well, but i added a warning that it is actually not specified https://github.com/european-epc-competence-center/rsa-rdfc-2025-cryptosuite

But for production software we will only use Ecdsa and Eddsa but i added it for consistency.

**Comment by msporny** [comment]
> is RSA out of scope for DataIntegrityProof?

RSA is deprecated for all intents and purposes for newer, modern systems and future use cases. It's a really old cryptography suite that is increasingly only supported for legacy use cases that cannot reasonably transition away from it. I would strongly recommend against building any new security system on top of RSA, and suggest ECDSA as your best bet in technology conservative organizations (such as government), and EdDSA if the customer has the flexibility to choose a better elliptic curve mechanism (though finding HSM support for it is more difficult than ECDSA).

All that said, since you've gone to the trouble of doing an RSA Cryptosuite implementation, and what you implemented seems to be valid... we /could/ standardize it if there was a strong driving reason to use it and a volunteer to move that specification through the standardization process.  At the very least, we should list it in the VC Extensions repository: 

https://www.w3.org/TR/vc-extensions/#securing-mechanisms

Any particular reason you chose RSA instead of ECDSA?

**Comment by F-Node-Karlsruhe** [comment]
We are on the same page there but we received the requirements for a wallet implementation from a customer, including support for PS256. While this is easy for VC-JWT it aint' for DI. Next we chose to provide each signed credential always as both JWT and DataIntegrityProof upon creation. VC-JWT will we widely supported for many EUDI use cases while DI will come in handy especially when signing product and supply chain data.

That's the way we've come and we included that all in our open source key-service implementation for reproducibility and trust.

I'm not sure if it is worth the effort to go ahead and specify the rsa-rdfc cryptosuite. Maybe if you confirm that it is just a bit of renaming to be done form ecdsa as they share a lot of spec. Meanwhile we'll continue to promote the more up to date cryptography :)
